### PR TITLE
Lock checkoutComplete, create order from adyen notification

### PR DESCRIPTION
**Pull request by korycins**
- refactor checkoutComplete mutation
- extract whole checkoutComplete logic to separate file
- use checkout_complete logic to create order from Adyen notification
- apply lock for handling the adyen notifciation and mutation checkoutComplete

### CODE REVIEW

**Review by fowczarek — APPROVED** [review]
For now it great.

**Comment by patrys** [comment]
Are the schema changes intentional?

**Comment by korycins** [comment]
> Are the schema changes intentional?
@patrys , I didn't add any changes for schema. Not sure why it raises the error. I assume that I need to update my branch with master

File: saleor/checkout/utils.py [hunk 1]
```diff
@@ -695,9 +696,14 @@ def abort_order_data(order_data: dict):
...
 
+def get_order(checkout_token: UUID) -> Optional[Order]:
+    """Get order based on checkout token."""
+    return Order.objects.confirmed().filter(checkout_token=checkout_token).first()
+
+
 @transaction.atomic
 def create_order(
-    *, checkout: Checkout, order_data: dict, user: User, redirect_url: str
+    *, checkout: Checkout, order_data: dict, user: User  # type: ignore
```
Comment (IKarbowiak) [inline saleor/checkout/utils.py:706]: Why there is `# type: ignore` in function definition?

**Review by IKarbowiak — APPROVED** [review]

File: saleor/checkout/utils.py (hunk 1 above)
Comment (fowczarek) [reply saleor/checkout/utils.py]: :policewoman:

Comment (korycins) [reply saleor/checkout/utils.py]: 👮‍♀️ 🚔

**Review by fowczarek — APPROVED** [review]
:rocket:

**Review by d-wysocki — APPROVED** [review]

File: saleor/checkout/complete_checkout.py [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+        gateway.payment_refund_or_void(payment)
+        raise ValidationError(
+            "Payment does not cover all checkout value.",
+            code=CheckoutErrorCode.CHECKOUT_NOT_FULLY_PAID.value,
+        )
+
+
+def get_order_data(checkout: models.Checkout, discounts: List[DiscountInfo]) -> dict:
+    try:
+        with transaction.atomic():
```
Comment (tomaszszymanski129) [inline saleor/checkout/complete_checkout.py:85]: is this atomic needed here?

File: saleor/payment/utils.py [hunk 3]
```diff
@@ -194,9 +194,9 @@ def get_already_processed_transaction_or_create_new_transaction(
     error_msg=None,
 ) -> Transaction:
     if gateway_response and gateway_response.transaction_already_processed:
-        transaction = get_already_processed_transaction(payment, gateway_response)
-        if transaction:
-            return transaction
+        txn = get_already_processed_transaction(payment, gateway_response)
+        if txn:
+            return txn
     return create_transaction(
```
Comment (tomaszszymanski129) [inline saleor/payment/utils.py:200]:
```suggestion
    return txn or create_transaction(
```

**Review by tomaszszymanski129 — APPROVED** [review]

**Comment by patrys** [comment]
13 new database queries, that's a lot.

**Comment by korycins** [comment]
> 13 new database queries, that's a lot.

@patrys, I will doube check them

File: saleor/checkout/complete_checkout.py (hunk 2 above)
Comment (korycins) [reply saleor/checkout/complete_checkout.py:85]: So what do you propose? It prepares the data for order like locking the vouchers etc

Comment (patrys) [reply saleor/checkout/complete_checkout.py:71]: Keep in mind that if we do locking, anything we `SELECT FOR UPDATE` will be unlocked as soon as the transaction terminates.

File: saleor/checkout/complete_checkout.py [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+from ..payment.models import Payment, Transaction
+from ..payment.utils import store_customer_id
+from . import models
+from .checkout_cleaner import clean_checkout_payment, clean_checkout_shipping
+from .error_codes import CheckoutErrorCode
+from .utils import abort_order_data, create_order, get_order, prepare_order_data
+
+
+@transaction.atomic
+def prepare_checkout(
```
Comment (maarcingebala) [inline saleor/checkout/complete_checkout.py:25]: I'm not sure what's the responsibility of this function, the name "prepare_checkout" doesn't explain much, there is also no docstring here.

File: saleor/checkout/complete_checkout.py (hunk 2 above)
Comment (maarcingebala) [reply saleor/checkout/complete_checkout.py:71]: I agree with @tomaszszymanski129, I would expect this function to only return some data. Also, `prepare_order_data` seems to be used only in "checkout complete", so why isn't it in this module?

File: saleor/checkout/complete_checkout.py [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+                payment=payment,
+                token=payment.token,
+                store_source=store_source,
+                additional_data=payment_data,
+            )
+        payment.refresh_from_db()
+        if not txn.is_success:
+            raise PaymentError(txn.error)
+    except PaymentError as e:
+        abort_order_data(order_data)
```
Comment (maarcingebala) [inline saleor/checkout/complete_checkout.py:110]: `abort_order_data` is also only used for completing checkouts, can we move it to this module? Also, this name is really awkward, we could fix it on this occasion.

File: saleor/checkout/complete_checkout.py [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+    discounts,
+    user,
+    tracking_code=None,
+    redirect_url=None,
+) -> Tuple[Optional[Order], bool, dict]:
+    action_required = False
+    action_data = {}  # type: ignore
+    order = get_order(checkout.token)
+    if order:
+        return order, action_required, action_data
```
Comment (maarcingebala) [inline saleor/checkout/complete_checkout.py:128]: What do you think about having a dataclass to represent the results of this function e.g. `CheckoutCompleteResult` @korycins  @patrys? It would be more elegant to process in places where the function is called, instead of expecting three tuple elements in this particular order? Just a suggestion, if you think it's overengineering, we can leave it as it is.

File: saleor/checkout/complete_checkout.py (hunk 2 above)
Comment (korycins) [reply saleor/checkout/complete_checkout.py:69]: @maarcingebala  @tomaszszymanski129  So what name do you propose?

Comment (tomaszszymanski129) [reply saleor/checkout/complete_checkout.py:69]: Hard to think of anything appropriate, maybe something like assemble_order? @maarcingebala any other idea?

File: saleor/checkout/complete_checkout.py (hunk 4 above)
Comment (korycins) [reply saleor/checkout/complete_checkout.py]: I've added the docstring. See if it is clear enough now

File: saleor/checkout/complete_checkout.py (hunk 5 above)
Comment (korycins) [reply saleor/checkout/complete_checkout.py:370]: Changed

File: saleor/checkout/complete_checkout.py (hunk 6 above)
Comment (korycins) [reply saleor/checkout/complete_checkout.py]: I would say it is overengineering but I can change it if you think that this is a good place for data class.

Comment (maarcingebala) [reply saleor/checkout/complete_checkout.py]: OK so let's leave it as it is now.

**Comment by korycins** [comment]
@patrys About the number of queries. After I've updated the benchmark tests, master has 145 queries for checkout complete. Here we have 119
```
test name                             left count      right count     duplicate count
complete checkout                                 145             119                  25
```

**Comment by patrys** [comment]
@korycins Was this branch rebased after the test change that increased the number of queries on master? Trying to make sure we're not comparing apples to oranges here.

**Comment by korycins** [comment]
@patrys Yes. I did that to get the current number of queries. - commit 3b9cafa

File: saleor/checkout/checkout_cleaner.py [hunk 7]
```diff
@@ -1,27 +1,28 @@
...
+    error_code: Union[Type[CheckoutErrorCode], Type[PaymentErrorCode]],
 ):
     if checkout.is_shipping_required():
         if not checkout.shipping_method:
             raise ValidationError(
                 {
                     "shipping_method": ValidationError(
                         "Shipping method is not set",
-                        code=error_code.SHIPPING_METHOD_NOT_SET,
+                        code=error_code.SHIPPING_METHOD_NOT_SET.value,
```
Comment (patrys) [inline saleor/checkout/checkout_cleaner.py:25]: Why are those changes needed?

File: saleor/core/transactions.py [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+
+
+@contextmanager
+def transaction_with_commit_on_errors():
+    """Perform transaction and raise an error in any occurred."""
+    error = None
+    with transaction.atomic():
+        try:
+            yield
+        except (DatabaseError, IntegrityError):
```
Comment (patrys) [inline saleor/core/transactions.py:13]: `IntegrityError` is a subclass of `DatabaseError`.

Comment (korycins) [reply saleor/core/transactions.py:13]: Yes, I know but I thought that passing it here explicitly will be a better way

**Review by maarcingebala — APPROVED** [review]

File: saleor/checkout/checkout_cleaner.py (hunk 7 above)
Comment (korycins) [reply saleor/checkout/checkout_cleaner.py:25]: I am getting: 
```
Unexpected type(s): (str, Union[CheckoutErrorCode, PaymentErrorCode]) Possible types: (Any, Optional[str]) (Any, Optional[str]) 
```
And then it raises the exception on mypy:
```
saleor/checkout/checkout_cleaner.py:43: error: Argument "code" to "ValidationError" has incompatible type "Union[CheckoutErrorCode, PaymentErrorCode]"; expected "Optional[str]"  [arg-type]
Found 1 error in 1 file (checked 1 source file)
```

File: saleor/core/transactions.py (hunk 8 above)
Comment (patrys) [reply saleor/core/transactions.py:13]: To me, it's confusing right now and caused me to double-check if there were maybe different classes with the same names.

Comment (korycins) [reply saleor/core/transactions.py:13]: Okay, I will remove `IntegrityError`

File: saleor/core/transactions.py [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+
+@contextmanager
+def transaction_with_commit_on_errors():
+    """Perform transaction and raise an error in any occurred."""
+    error = None
+    with transaction.atomic():
+        try:
+            yield
+        except DatabaseError:
+            raise
```
Comment (IKarbowiak) [inline saleor/core/transactions.py:14]: Shouldn't we raise something here?

Comment (korycins) [reply saleor/core/transactions.py:14]: We re-raise the DatabaseError here

**Review by IKarbowiak — APPROVED** [review]

**Review by fowczarek — APPROVED** [review]
